addons/website/static/src/components/dialog/edit_menu.js

```diff
@@ -122,40 +121,75 @@ export class EditMenuDialog extends Component {
-                expression: '()(.+)', // nestedSortable takes the second match of an expression (*sigh*)
-                isAllowed: (placeholder, placeholderParent, currentItem) => {
-                    return !placeholderParent
-                        || !currentItem[0].dataset.isMegaMenu && !placeholderParent[0].dataset.isMegaMenu;
-                },
-            });
+        useNestedSortable({
+            ref: this.menuEditor,
+            handle: "div",
+            nest: true,
+            maxLevels: 2,
+            onDrop: this._moveMenu.bind(this),
+            isAllowed: this._isAllowedMove.bind(this),
+            useElementSize: true,
         });
     }
 
+    get map() {
+        return this.state.flatMenus;
+    }
+
     populate(map, menu) {
         map.set(menu.fields['id'], menu);
         for (const submenu of menu.children) {
             this.populate(map, submenu);
         }
     }
 
+    _isAllowedMove(current, elementSelector) {
+        const currentIsMegaMenu = current.element.dataset.isMegaMenu === "true"
+        if (!currentIsMegaMenu) {
+            return current.placeHolder.parentNode.closest(`${elementSelector}[data-is-mega-menu="true"]`) === null;
+        }
+        const isDropOnRoot = current.placeHolder.parentNode.closest(elementSelector) === null;
+        return currentIsMegaMenu && isDropOnRoot;
+    }
+
+    _getMenuIdForElement(element) {
+        const menuIdStr = element.dataset.menuId;
+        const menuId = parseInt(menuIdStr);
+        return isNaN(menuId) ? menuIdStr : menuId;
+    }
+
+    _moveMenu({ element, parent, previous, next }) {
+        const menuId = this._getMenuIdForElement(element);
+        const menu = this.map.get(menuId);
+
+        // Remove element from parent's children (since we are moving it, this is the mandatory first step)
+        const parentId = menu.fields['parent_id'] || this.state.rootMenu.fields['id'];
+        let parentMenu = this.map.get(parentId);
+        parentMenu.children = parentMenu.children.filter((m) => m.fields['id'] !== menuId);
+
+        // Determine next parent
+        const menuParentId = parent ? this._getMenuIdForElement(parent.closest("li")) : this.state.rootMenu.fields['id'];
+        parentMenu = this.map.get(menuParentId);
+        menu.fields['parent_id'] = parentMenu.fields['id'];
+
+        // Determine at which position we should place the element
+        if (previous) {
+            previous = this.map.get(this._getMenuIdForElement(previous));
+            const index = parentMenu.children.findIndex((child) => child === previous);
+            parentMenu.children.splice(index + 1, 0, menu);
+        } else if (next) {
+            next = this.map.get(this._getMenuIdForElement(next));
+            const index = parentMenu.children.findIndex((child) => child === next);
+            parentMenu.children.splice(index, 0, menu);
+        } else {
+            parentMenu.children.push(menu);
+        }
```

Comment:
```suggestion
        if (previous) {
            previous = this.map.get(this._getMenuIdForElement(previous));
            const index = parentMenu.children.findIndex((child) => child === previous);
            parentMenu.children.splice(index + 1, 0, menu);
        } else {
            parentMenu.children.unshift(menu);
        }
```
If it doesn't have a previous, it means it's the first right? 😄 Whether it has a next element shouldn't matter I think.